log4j-docgen/src/main/java/org/apache/logging/log4j/docgen/processor/DocGenProcessor.java

```diff
@@ -0,0 +1,692 @@
+        // First step: document plugins
+        roundEnv.getElementsAnnotatedWith(Plugin.class).forEach(element -> {
+            if (element instanceof final TypeElement typeElement) {
+                final PluginType pluginType = new PluginType();
+                final Plugin pluginAnnotation = element.getAnnotation(Plugin.class);
+                processPlugin(typeElement, pluginAnnotation, pluginType);
+                pluginTypes.put(pluginType.getClassName(), pluginType);
+            }
+        });
+        pluginTypes.values().forEach(set::addPlugin);
+        // Second step: document abstract types
+        abstractTypes.values().forEach(type -> {
+            final AbstractType abstractType = new AbstractType();
+            processAbstractType(type, abstractType);
+            if (!abstractType.getDescription().getText().isEmpty()) {
+                set.addAbstractType(abstractType);
+            }
+        });
+        // Second step: document scalars
+        scalarTypes.values().forEach(type -> {
+            final ScalarType scalarType = new ScalarType();
+            processScalarType(type, scalarType);
+            set.addScalar(scalarType);
+        });
+        // Write the result file
+        if (roundEnv.processingOver()) {
+            try {
+                final FileObject output = processingEnv
+                        .getFiler()
+                        .createResource(StandardLocation.CLASS_OUTPUT, "", "META-INF/log4j/plugins.xml");
+
+                try (final Writer writer = output.openWriter()) {
+                    new PluginBundleStaxWriter().write(writer, set);
+                }
+            } catch (final IOException | XMLStreamException e) {
+                messager.printMessage(
+                        Diagnostic.Kind.ERROR,
+                        "An error occurred while writing to `META-INF/log4j/plugins.xml`: " + e.getMessage());
+            }
+        }
+        return false;
+    }
+
+    private void processType(final QualifiedNameable element, final Type docgenType) {
+        // Class name
+        docgenType.setClassName(element.getQualifiedName().toString());
+        // Description
+        docgenType.setDescription(createDescription(element, null));
+    }
+
+    private void processAbstractType(final QualifiedNameable element, final AbstractType abstractType) {
+        processType(element, abstractType);
+    }
+
+    private void processScalarType(final TypeElement element, final ScalarType scalarType) {
+        processType(element, scalarType);
+        if (types.isSubtype(element.asType(), enumType)) {
+            for (final Element member : element.getEnclosedElements()) {
+                if (member instanceof final VariableElement field
+                        && field.getModifiers().contains(Modifier.STATIC)
+                        && types.isSameType(field.asType(), element.asType())) {
+                    final ScalarValue value = new ScalarValue();
```

Comment:
@ppkarwasz, around this method I started losing my marbles. I navigated the branch locally in my IDE too, it did not help either. Hence, there is nothing much I can do than assuming _"it works"_.

Reply:
Maybe the code lacks some comments. Now it should be better.

Thank your for the review.